src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,105 @@
 package core.basesyntax;
 
 public class MyHashMap<K, V> implements MyMap<K, V> {
+    private static final int INITIAL_CAPACITY = 16;
+    private static final float LOAD_FACTOR = 0.75f;
+    private Entry<K, V>[] table;
+    private int size;
+    private Entry<K, V> nullKeyEntry;
+
+    @SuppressWarnings("unchecked")
+    public MyHashMap() {
+        table = (Entry<K, V>[]) new Entry[INITIAL_CAPACITY];
+    }
 
     @Override
     public void put(K key, V value) {
+        if (key == null) {
+            if (nullKeyEntry == null) {
+                nullKeyEntry = new Entry<>(null, value,null);
+                size++;
+            } else {
+                nullKeyEntry.value = value;
+            }
+            return;
+        }
+        int hash = hash(key);
+        int index = indexFor(hash, table.length);
 
+        for (Entry<K, V> e = table[index]; e != null; e = e.next) {
+            if (e.key.equals(key)) {
+                e.value = value;
+                return;
+            }
+        }
+
+        addEntry(hash, key, value,index);
     }
 
     @Override
     public V getValue(K key) {
+        if (key == null) {
+            return nullKeyEntry == null ? null : nullKeyEntry.value;
+        }
+        int hash = hash(key);
+        int index = indexFor(hash, table.length);
+        for (Entry<K, V> e = table[index]; e != null; e = e.next) {
+            if (e.key.equals(key)) {
```

Comment:
replace two occurrences of this comparison logic with null safe comparison like in Objects.equals() and you won't need special handling of null and  private Entry<K, V> nullKeyEntry field